### ISSUE TITLE
Path must be normalized

### ISSUE THREAD

**Issue by afpatmin** [dart-lang/build#2020]
OS: Windows 10
Dart SDK: 2.2.0-dev.1.1
build_runner: 1.2.5
build_web_compilers: 1.1.0

Running `webdev serve` in my project crashes with the following output:

```
PS C:\Users\p_min\Documents\Workspace\Clients\Boardon\boardon.io\client> webdev serve --live-reload
Creating build script, took 8858ms
[INFO] Setting up file watchers completed, took 15ms
[INFO] Waiting for all file watchers to be ready completed, took 132ms
[INFO] Building new asset graph completed, took 2.1s
[INFO] Checking for unexpected pre-existing outputs. completed, took 1ms
[SEVERE] angular on web/main.dart:

Invalid argument(s): Path must be normalized : /onboarding_client\web/main.dart
[SEVERE] angular on lib/app_component.dart:

Invalid argument(s): Path must be normalized : /onboarding_client\lib/app_component.dart
[SEVERE] angular on package:angular/angular.dart:

Invalid argument(s): Path must be normalized : /angular\lib/angular.dart
[SEVERE] angular on package:angular/core.dart:

Invalid argument(s): Path must be normalized : /angular\lib/core.dart
[SEVERE] angular on package:angular/di.dart:

Invalid argument(s): Path must be normalized : /angular\lib/di.dart
[SEVERE] angular on package:angular/experimental.dart:

Invalid argument(s): Path must be normalized : /angular\lib/experimental.dart
[SEVERE] angular on package:angular/meta.dart:

Invalid argument(s): Path must be normalized : /angular\lib/meta.dart
[SEVERE] angular on package:angular/security.dart:

Invalid argument(s): Path must be normalized : /angular\lib/security.dart
[SEVERE] angular on package:angular/src/bootstrap/modules.dart:

...
```
The errors started only after running pub upgrade today.

I'm guessing it has to do with the referenced paths containing both '\\' and '/' delimiters? I've checked my project's imports just in case and they all have consistent '/' delimiters.

**Comment by jakemac53** [comment]
This is likely related to the build_resolvers update which was released recently, cc @natebosch

**Comment by jakemac53** [comment]
@afpatmin can you run with `--verbose` and hopefully we can get a stack trace to help diagnose the issue.

**Comment by afpatmin** [comment]
Sure, here are the first two [SEVERE] outputs, let me know if you need any more

```
...
[FINEST] Watch:Got modify event for "asset:onboarding_client/.dart_tool/build/generated/onboarding_client/lib/app_component.css in c:\users\p_min\documents\workspace\clients\boardon\boardon.io\client
[SEVERE] angular on web/main.dart:

Invalid argument(s): Path must be normalized : /onboarding_client\web/main.dart
package:analyzer/file_system/memory_file_system.dart 246:7                                    MemoryResourceProvider._ensureAbsoluteAndNormalized
package:analyzer/file_system/memory_file_system.dart 151:5                                    MemoryResourceProvider.newFile
package:build_resolvers/src/build_asset_uri_resolver.dart 45:30                               BuildAssetUriResolver.performResolve.processAsset.<fn>
dart:async                                                                                    _completeOnAsyncReturn
package:pool/pool.dart                                                                        Pool.withResource
package:build_runner_core/src/asset/file_based.dart 34:39                                     FileBasedAssetReader.readAsBytes.<fn>
dart:async                                                                                    _completeOnAsyncReturn
package:pool/pool.dart                                                                        Pool.withResource
package:build_runner_core/src/asset/file_based.dart 133:26                                    _fileForOrThrow
package:build_runner_core/src/asset/file_based.dart 33:48                                     FileBasedAssetReader.readAsBytes
package:build_runner_core/src/asset/build_cache.dart 39:17                                    BuildCacheReader.readAsBytes
package:build_runner_core/src/asset/cache.dart 74:25                                          CachingAssetReader.readAsBytes.<fn>
dart:collection                                                                               __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.putIfAbsent
package:build_runner_core/src/asset/cache.dart 72:38                                          CachingAssetReader.readAsBytes
package:build_runner_core/src/asset/cache.dart 96:15                                          CachingAssetReader.readAsString.<fn>
dart:collection                                                                               __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.putIfAbsent
package:build_runner_core/src/asset/cache.dart 94:39                                          CachingAssetReader.readAsString
package:build_runner_core/src/asset/reader.dart 114:28                                        SingleStepReader.readAsString.<fn>.<fn>
package:build_runner_core/src/util/async.dart 13:20                                           doAfter
package:build_runner_core/src/asset/reader.dart 113:14                                        SingleStepReader.readAsString.<fn>
package:build_runner_core/src/util/async.dart 13:20                                           doAfter
package:build_runner_core/src/asset/reader.dart 109:21                                        SingleStepReader.readAsString
package:build                                                                                 BuildStepImpl.readAsString
package:build_resolvers/src/build_asset_uri_resolver.dart 38:30                               BuildAssetUriResolver.performResolve.processAsset
dart:core                                                                                     List.forEach
package:build_resolvers/src/build_asset_uri_resolver.dart 71:17                               BuildAssetUriResolver.performResolve
package:build_resolvers/src/resolver.dart 145:24                                              AnalyzerResolvers.get
package:build_runner_core/src/performance_tracking/performance_tracking_resolvers.dart 19:58  PerformanceTrackingResolvers.get.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 303:15                        _NoOpBuilderActionTracker.trackStage
package:build_runner_core/src/performance_tracking/performance_tracking_resolvers.dart 19:16  PerformanceTrackingResolvers.get
package:build                                                                                 BuildStepImpl.resolver
package:source_gen/src/builder.dart 63:30                                                     _Builder.build
package:build                                                                                 runBuilder
package:build_runner_core/src/generate/build_impl.dart 460:19                                 _SingleBuild._runForInput.<fn>.<fn>.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 303:15                        _NoOpBuilderActionTracker.trackStage
package:build_runner_core/src/generate/build_impl.dart 458:23                                 _SingleBuild._runForInput.<fn>.<fn>
package:timing/src/timing.dart 222:44                                                         NoOpTimeTracker.track
package:build_runner_core/src/generate/build_impl.dart 417:22                                 _SingleBuild._runForInput.<fn>
package:pool/pool.dart 127:28                                                                 Pool.withResource
package:build_runner_core/src/generate/build_impl.dart 413:17                                 _SingleBuild._runForInput
package:build_runner_core/src/generate/build_impl.dart 387:14                                 _SingleBuild._runLazyPhaseForInput.<fn>
dart:collection                                                                               __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.putIfAbsent
package:build_runner_core/src/generate/build_impl.dart 371:24                                 _SingleBuild._runLazyPhaseForInput
package:build_runner_core/src/generate/build_impl.dart 345:17                                 _SingleBuild._matchingPrimaryInputs.<fn>
dart:async                                                                                    Future.wait
package:build_runner_core/src/generate/build_impl.dart 336:18                                 _SingleBuild._matchingPrimaryInputs
dart:async                                                                                    _AsyncAwaitCompleter.start
package:build_runner_core/src/generate/build_impl.dart 332:46                                 _SingleBuild._matchingPrimaryInputs
package:build_runner_core/src/generate/build_impl.dart 308:23                                 _SingleBuild._runPhases.<fn>.<fn>
dart:async                                                                                    _AsyncAwaitCompleter.start
package:build_runner_core/src/generate/build_impl.dart 305:70                                 _SingleBuild._runPhases.<fn>.<fn>
package:build_runner_core/src/generate/performance_tracker.dart 187:15                        _NoOpBuildPerformanceTracker.trackBuildPhase
package:build_runner_core/src/generate/build_impl.dart 305:47                                 _SingleBuild._runPhases.<fn>
dart:async                                                                                    _completeOnAsyncReturn
package:build_runner_core/src/generate/build_impl.dart                                        _SingleBuild._runBuilder
dart:async                                                                                    _completeOnAsyncReturn
package:pool/pool.dart                                                                        Pool.withResource
dart:async                                                                                    _completeOnAsyncReturn
package:build_runner_core/src/generate/build_impl.dart                                        _SingleBuild._runForInput.<fn>.<fn>
dart:async                                                                                    _completeOnAsyncReturn
package:build_runner_core/src/generate/build_impl.dart                                        _SingleBuild._setOutputsState
dart:async                                                                                    _completeOnAsyncReturn
package:pool/pool.dart                                                                        Pool.withResource

[FINEST] Watch:Got remove event for "asset:onboarding_client/.dart_tool/build/098d3ee73e6cc294616d3a2e2c3c81ad/error_cache/onboarding_client/934/web/main.dart in c:\users\p_min\documents\workspace\clients\boardon\boardon.io\client
[SEVERE] angular on lib/app_component.dart:

Invalid argument(s): Path must be normalized : /onboarding_client\lib/app_component.dart
package:analyzer/file_system/memory_file_system.dart 246:7       MemoryResourceProvider._ensureAbsoluteAndNormalized
package:analyzer/file_system/memory_file_system.dart 151:5       MemoryResourceProvider.newFile
package:build_resolvers/src/build_asset_uri_resolver.dart 45:30  BuildAssetUriResolver.performResolve.processAsset.<fn>
...
```

**Comment by jakemac53** [comment]
Note that the tests are also failing on windows due to this - we just don't run these tests on windows (only e2e tests, none of which apparently use a resolver).

For now I will also add the build_resolvers tests to the windows bots.

**Comment by jakemac53** [comment]
hmmm, I am struggling to come up with a fix the analyzer is happy with here, @scheglov can you help?

I tried wrapping [this line](https://github.com/dart-lang/build/blob/5dbee7766f85499591c5abfd6bf5e3d1a857e849/build_resolvers/lib/src/build_asset_uri_resolver.dart#L108) in a `p.normalize` but that seems to give other problems when running the tests (on windows at least):

**Comment by scheglov** [comment]
What path style do you expect the `MemoryResourceProvider` to use on Windows?
Posix or Windows?
The rest of the `build` package need to ensure that this path style is used.

`/a\lib/a.dart` is patently invalid, it is neither one, not another.

Usually solution is to use the the `pathContext` of the `ResourceProvider` to compose paths, e.g. using `join()`.

**Comment by jakemac53** [comment]
> What path style do you expect the `MemoryResourceProvider` to use on Windows?
> Posix or Windows?

If possible, we would use Posix style paths always (the build system internally always does this, so that serialized asset paths are cross-platform compatible). I did try just doing `p.url.join` here instead of `p.join`, but the I get an error from the analyzer because it asserts that `p.normalize` on the path returns the original path.

> `/a\lib/a.dart` is patently invalid, it is neither one, not another.

Right, the `p.normalize` changed it to the current OS version (I for got to mention I also changed to use `p.url.join` so that `p.normalize` could work properly).

> Usually solution is to use the the `pathContext` of the `ResourceProvider` to compose paths, e.g. using `join()`.

Ah, that gives us a potential solution of doing the `p.url.join` and then setting the `pathContext` of our `ResourceProvider` to always be a Posix context so it doesn't complain about normalization. I will try that and follow up here.

**Comment by jakemac53** [comment]
Sweet, overriding the context to always be a Posix context solves things, thanks @scheglov for the pointers.

**Comment by scheglov** [comment]
I would not use `p.url`.
It might be using the same path separator, but it is not `posix`.

I would create `MemoryResourceProvider` with `posix` style as well.

**Comment by jakemac53** [comment]
The goal in the build system was to use the most platform agnostic thing - which would be url style paths we determined, hence we chose `p.url`. But I really don't think it matters as long as we are consistent 🤷‍♂️

**Comment by jakemac53** [comment]
heh @scheglov not sure why but the tests fail if I provide `p.url` to the MemoryResourceProvider, not sure exactly the difference there (but it only appears to matter to the analyzer internally).

For now I will just switch to `p.posix` for this code path to get windows unblocked, and ill file an issue to try and determine what is going on.

**Comment by afpatmin** [comment]
I have come across another issue #2022 - but on linux this time. Is it by any chance related?

**Comment by jakemac53** [comment]
@afpatmin Seems likely related to the same change yes, can you try both with the current fix I have out?

You can depend on the `fix-windows` branch of this repro for the build_resolvers package, like this in your pubspec.yaml:

```
dependency_overrides:
  build_resolvers:
    git:
      url: https://github.com/dart-lang/build.git
      ref: fix-windows
      path: build_resolvers
```

**Comment by afpatmin** [comment]
@jakemac53 
Yep that solved the first issue, but not the #2022 one unfortunately.

**Comment by afpatmin** [comment]
Nice, thanks for helping out. 

You're right I have analyzer: ^0.35.0 specified in dependency overrides.

**Comment by jakemac53** [comment]
> You're right I have analyzer: ^0.35.0 specified in dependency overrides.

Ok cool, just wanted to make sure that was what was happening and this wasn't happening for people more broadly.

> Edit: Just tested setting it to ^0.34.0, same issue

Right, so this is a bit confusing because analyzer isn't actually the issue, its a build_resolvers issue (you could override that back to the previous version to get unblocked). It is just that the latest build_resolvers requires the latest analyzer.

The versioning here is confusing since most users don't depend on build_resolvers directly, but it had a breaking change in that it now uses different analyzer apis internally which work differently. So, we sort of abused the analyzer version in order to roll it out in a way that we shouldn't break people (and to users this will actually surface as what appears like breaking changes in the analyzer).

**Comment by jakemac53** [comment]
Published the first fix as build_resolvers 1.0.1
